plugins/usbdmx/AsyncUsbTransceiverBase.cpp

```diff
@@ -29,36 +31,69 @@ namespace usbdmx {
 
 using ola::usb::LibUsbAdaptor;
 
-namespace {
-
 /*
  * Called by libusb when the transfer completes.
  */
 #ifdef _WIN32
 __attribute__((__stdcall__))
 #endif  // _WIN32
-void AsyncCallback(struct libusb_transfer *transfer) {
+void AsyncTransferCallback(struct libusb_transfer *transfer) {
   AsyncUsbTransceiverBase *widget = reinterpret_cast<AsyncUsbTransceiverBase*>(
     transfer->user_data);
-  widget->TransferComplete(transfer);
+  widget->TransferCompleteInternal(transfer);
+}
+
+void AsyncUsbTransceiverBase::TransferCompleteInternal(
+  struct libusb_transfer *transfer) {
+  {
+    ola::thread::MutexLocker locker(&m_mutex);
+    inflight_set::iterator i = m_inflight.find(transfer);
+    if (i == m_inflight.end()) {  // not found
+      OLA_WARN << "Mismatched libusb transfer: "
+               << transfer << " not in inflight set";
+      return;
+    }
+
+    m_inflight.erase(i);
+    m_idle.push(transfer);
+  }
+
+  if (transfer->status == LIBUSB_TRANSFER_NO_DEVICE) {
+    m_device_disconnected = true;
+  }
+
+  TransferComplete(transfer);
+
+  {
+    ola::thread::MutexLocker locker(&m_mutex);
+    if (m_inflight.empty())
+      m_cond.Signal();
```

Comment:
Can we have some { } again

Reply:
I don't get this obsession with `{ }`, but sure. GCC has a warning now for when a dodgy multi-statement macro is expanded in a bracket-less `if` you know ;)

Reply:
Just I've been tripped up by writing something like this a few times in the past:
```suggestion
    if (m_inflight.empty())
      m_cond.Signal();
      OLA_DEBUG << "Inflight empty";
```

Would GCC catch that too now?

Reply:
Yup, that's `-Wmisleading-indentation` which is part of `-Wall`:

```
void foo() {}
int main() {
        if (true)
                foo();
                foo();
}
```

`gcc -Wall test.cc`:
```
test.cc: In function ‘int main()’:
test.cc:3:9: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
         if (true)
         ^~
test.cc:5:17: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
                 foo();
```

There's also `-Wmultistatement-macros`, also part of `-Wall` which is the case I meant:

```
void foo() {}
#define dodgy() foo(); foo();
int main() {
        if (true)
                dodgy()
}
```

This will give:

```
test.cc: In function ‘int main()’:
test.cc:2:17: warning: macro expands to multiple statements [-Wmultistatement-macros]
 #define dodgy() foo(); foo();
                 ^~~
test.cc:5:17: note: in expansion of macro ‘dodgy’
                 dodgy()
                 ^~~~~
test.cc:4:9: note: some parts of macro expansion are not guarded by this ‘if’ clause
         if (true)
         ^~
```

Reply:
Ah very fancy. I'd probably still favour the brackets too though, as they don't really hurt.

Reply:
Done.